File: backend/src/schemas/task.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime, date
from uuid import UUID
from .user import UserProfile


class TaskBase(BaseModel):
    title: str
    description: Optional[str] = None
    status: Optional[str] = "pending"
    priority: str
    due_date: Optional[datetime] = None
    tags: Optional[str] = None  # Comma-separated tags
    is_recurring: bool = False
    recurrence_pattern: Optional[dict] = None  # Recurrence configuration


class TaskCreate(TaskBase):
    title: str
    priority: str
    user_id: Optional[UUID] = None


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    priority: Optional[str] = None
    due_date: Optional[datetime] = None
    tags: Optional[str] = None  # Comma-separated tags
    is_recurring: Optional[bool] = None
    recurrence_pattern: Optional[dict] = None  # Recurrence configuration
    completed_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class TaskResponse(TaskBase):
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None
    is_recurring: bool = False


class TaskFilter(BaseModel):
    status: Optional[str] = None
    priority: Optional[str] = None
    due_date_from: Optional[datetime] = None
    due_date_to: Optional[datetime] = None
    search: Optional[str] = None
    completed: Optional[bool] = None


class TaskListResponse(BaseModel):
    tasks: List[TaskResponse]
    total: int


class RecurrencePatternBase(BaseModel):
    pattern_type: str
    interval: int = 1
    end_condition_type: str
    end_count: Optional[int] = None
    end_date: Optional[date] = None


class RecurrencePatternCreate(RecurrencePatternBase):
    user_id: Optional[UUID] = None


class RecurrencePatternResponse(RecurrencePatternBase):
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/task.ts
// Task interface for frontend
export interface Task {
  id: string;
  title: string;
  description?: string;
  status: 'pending' | 'in-progress' | 'completed' | 'cancelled' | 'on_hold';
  priority: 'low' | 'medium' | 'high';
  tags?: string;  // Comma-separated string
  dueDate?: string; // ISO date string
  completed: boolean;
  completedAt?: string; // ISO date string
  recurrencePattern?: RecurrencePattern;
  userId: string;
  createdAt: string; // ISO date string
  updatedAt: string; // ISO date string
}

export interface RecurrencePattern {
  type: 'daily' | 'weekly' | 'monthly' | 'yearly';
  interval: number;
  end_date?: string; // ISO date string
  occurrences?: number;
  days_of_week?: number[]; // 0-6 where 0 is Sunday
  days_of_month?: number[]; // 1-31
}

// API Request/Response types
export interface CreateTaskRequest {
  title: string;
  description?: string;
  status?: 'pending' | 'in-progress' | 'completed' | 'cancelled' | 'on_hold';
  priority?: 'low' | 'medium' | 'high';
  tags?: string;
  dueDate?: string;
  recurrencePattern?: RecurrencePattern;
}

export interface UpdateTaskRequest {
  title?: string;
  description?: string;
  status?: 'pending' | 'in-progress' | 'completed' | 'cancelled' | 'on_hold';
  priority?: 'low' | 'medium' | 'high';
  tags?: string;
  dueDate?: string;
  recurrencePattern?: RecurrencePattern;
}

export interface TaskFilter {
  status?: 'pending' | 'in-progress' | 'completed' | 'cancelled' | 'on_hold';
  priority?: 'low' | 'medium' | 'high';
  tags?: string;
  has_recurrence?: boolean;
  due_date_start?: string;
  due_date_end?: string;
  completed?: boolean;
  search?: string;
  skip?: number;
  limit?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface ToggleCompleteRequest {
  completed: boolean;
}

export interface ErrorResponse {
  error: {
    code: string;
    message: string;
    details?: string;
  };
}

// User interface
export interface User {
  id: string;
  email: string;
  name: string;  // Make name required
  createdAt: string;
  updatedAt: string;
  is_active?: boolean;
}

// API Response types
export interface TaskResponse extends Task {}
export interface UserResponse extends User {}
export interface TokenResponse {
  access_token: string;
  token_type: string;
  refresh_token: string;
  user: User;
}

export interface RefreshTokenRequest {
  refresh_token: string;
}

export interface RefreshTokenResponse {
  access_token: string;
  token_type: string;
}